staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go

```diff
@@ -107,22 +123,51 @@ func (s *specProxier) AddUpdateAPIService(handler http.Handler, apiservice *v1.A
 	}
 }
 
+func getGroupVersionStringFromAPIService(apiService v1.APIService) string {
+	if apiService.Spec.Group == "" && apiService.Spec.Version == "" {
+		return ""
+	}
+	return "apis/" + apiService.Spec.Group + "/" + apiService.Spec.Version
+}
+
 // UpdateAPIServiceSpec updates all the OpenAPI v3 specs that the APIService serves.
 // It is thread safe.
 func (s *specProxier) UpdateAPIServiceSpec(apiServiceName string) error {
 	s.rwMutex.Lock()
 	defer s.rwMutex.Unlock()
+	return s.updateAPIServiceSpecLocked(apiServiceName)
+}
 
+func (s *specProxier) updateAPIServiceSpecLocked(apiServiceName string) error {
 	apiService, exists := s.apiServiceInfo[apiServiceName]
 	if !exists {
 		return fmt.Errorf("APIService %s does not exist for update", apiServiceName)
 	}
 
-	gv, err := s.downloader.OpenAPIV3Root(apiService.handler)
+	if !apiService.isLegacyAPIService {
```

Comment:
Does this mean that it can go from not legacy to legacy but never the other way around?

Reply:
After looking into it again, I think having a permanent flag may not make sense.

An APIService either serves openapi v3 or does not serve openapi v3 (legacy). We check that by querying the v3 endpoint and seeing whether a 404 not found is returned.

If an APIService returns 404, I'm not sure if we can fully claim that this APIService will not serve OpenAPI V3 forever. It could be temporarily unavailable, or it could be upgraded to serve OpenAPI V3. We would still need to ping the v3 endpoint to verify these cases.